src/shared/treeview/resource.ts

```diff
@@ -17,37 +23,234 @@ export interface Resource {
+            return this.getChangedEmitter().event
+        }
+    }
+
+    public getTreeItem(): vscode.TreeItem | Promise<vscode.TreeItem> {
+        const collapsibleState =
+            this.options?.childrenProvider !== undefined
                 ? vscode.TreeItemCollapsibleState.Collapsed
                 : vscode.TreeItemCollapsibleState.None
 
-        const item = new vscode.TreeItem(this.content.label, collapsed)
-        assign(this.content, item)
+        // The two branches are for tree shim optimizations
+        const item = this.resource.getTreeItem()
+        if (item instanceof Promise) {
+            return item.then(i => {
+                i.collapsibleState = collapsibleState
+                return i
+            })
+        }
+        item.collapsibleState = collapsibleState
 
         return item
     }
+
+    public async getChildren(): Promise<TreeNode<U>[] | [...TreeNode<U>[], TreeNode]> {
+        if (!this.options?.childrenProvider) {
+            return []
+        }
+
+        const { placeholder, childrenProvider, onError } = this.options
+        const children = this.fromProvider(childrenProvider)
+        const [succeeded, result] = await handleError(children, onError ?? createErrorItem)
+
+        if (result.length === 0 && placeholder !== undefined) {
+            return [typeof placeholder === 'string' ? createPlaceholderItem(placeholder) : placeholder]
+        }
+
+        if (succeeded) {
+            if (this.options.sort) {
+                result.sort(this.options.sort.bind(this.options))
+            }
+
+            if (this.loader && !this.loader.done) {
+                return this.withLoadMoreNode(this.loader, result)
+            }
+        }
+
+        return result
+    }
+
+    public dispose(): void {
+        this.loader?.dispose()
+        vscode.Disposable.from(...this.disposables).dispose()
+    }
+
+    private refresh(): void {
+        this.loader?.dispose()
+        this.loader = undefined
+        this.getChangedEmitter().fire()
+    }
+
+    private async fromProvider(provider: ResourceProvider<TreeNode<U>>): Promise<TreeNode<U>[]> {
+        if (this.loader) {
+            return this.loadFrom(this.loader)
+        }
+
+        const children = provider.listResources()
+        if (isAsyncIterable(children)) {
+            const loader = (this.loader ??= new PageLoader(children))
+
+            return this.loadFrom(loader)
+        }
+
+        return children
+    }
+
+    private async loadFrom(loader: PageLoader<TreeNode<U>>): Promise<TreeNode<U>[]> {
+        const pages = await loader.get()
+        const merged = pages.reduce((a, b) => a.concat(b), [])
+
+        return merged
+    }
+
+    private withLoadMoreNode(loader: PageLoader<TreeNode<U>>, result: TreeNode<U>[]): [...typeof result, TreeNode] {
+        // TODO: optimize this by making `loadMore` apart of the `ResourceTreeNode` prototype?
+        // but then we can't make the method private :/
```

Comment:
why is private important? if we think there is a risk of caller mistakes, we could add a lint rule in the future.

Reply:
Was mainly because I had the data source decoupled from the consumer, but I might as well combine them.